tactics/tactics.ml

```diff
@@ -2863,17 +2863,22 @@ let generalize_dep ?(with_let=false) c =
       | _ -> tothin
   in
   let cl' = it_mkNamedProd_or_LetIn (pf_concl gl) to_quantify in
-  let body =
-    if with_let then
-      match EConstr.kind sigma c with
-      | Var id -> id |> (fun id -> pf_get_hyp id gl) |> NamedDecl.get_value
-      | _ -> None
-    else None
+  let is_var, body = match EConstr.kind sigma c with
+  | Var id ->
+    let body = id |> (fun id -> pf_get_hyp id gl) |> NamedDecl.get_value in
+    let is_var = Option.is_empty body && not (List.mem id init_ids) in
+    if with_let then is_var, body
+    else is_var, None
```

Comment:
Suggestion (instead of the three lines above):

~~~
Option.is_empty body && not (List.mem id init_ids),
if with_let then body else None
~~~

Reply:
I'm a crypto-C developer, I don't like to abuse the expression-based semantics...

Reply:
No problem. Thanks.